File: e2e/tests/steps/connection.rs
use cucumber::{then, when};

use crate::steps::parse_media_kind;
use crate::World;

#[then(regex = r"^(\S+) receives connection with (\S+)$")]
async fn then_member_receives_connection(
    world: &mut World,
    id: String,
    responder_id: String,
) {
    let member = world.get_member(&id).unwrap();
    member
        .connections()
        .wait_for_connection(responder_id.clone())
        .await
        .unwrap();
}

#[then(regex = r"^(\S+) doesn't receive connection with (\S+)$")]
async fn then_member_doesnt_receive_connection(
    world: &mut World,
    id: String,
    responder_id: String,
) {
    let member = world.get_member(&id).unwrap();
    assert!(member
        .connections()
        .get(responder_id)
        .await
        .unwrap()
        .is_none());
}

#[then(regex = r"^(\S+)'s connection with (\S+) closes$")]
async fn then_connection_closes(
    world: &mut World,
    id: String,
    partner_id: String,
) {
    let member = world.get_member(&id).unwrap();
    let connection =
        member.connections().get(partner_id).await.unwrap().unwrap();
    connection.wait_for_close().await.unwrap();
}

#[when(regex = r"^(\S+) (enables|disables) (audio|video) receive from (\S+)")]
async fn when_connection_changes_remote_media_state(
    world: &mut World,
    id: String,
    action: String,
    kind: String,
    partner_id: String,
) {
    let kind = parse_media_kind(&kind).unwrap();

    let member = world.get_member(&id).unwrap();
    let connection =
        member.connections().get(partner_id).await.unwrap().unwrap();

    if action.contains("enables") {
        connection.enable_remote_media(kind).await.unwrap();
    } else {
        connection.disable_remote_media(kind).await.unwrap();
    }
}

File: flutter/test/e2e/steps/connection.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gherkin/gherkin.dart';

import 'package:medea_jason/medea_jason.dart';
import '../world/custom_world.dart';
import '../world/more_args.dart';

const bool isSfu = bool.fromEnvironment('SFU', defaultValue: false);

List<StepDefinitionGeneric> steps() {
  return [
    thenConnectionCloses,
    thenMemberReceivesConnection,
    thenMemberConnectionStateIs,
    whenConnectionChangesRemoteMediaState,
    whenMemberReceivesConnection,
    thenMemberDoesntReceiveConnection,
  ];
}

StepDefinitionGeneric thenConnectionCloses = then2<String, String, CustomWorld>(
  r"(\S+)'s connection with (\S+) closes$",
  (id, partnerId, context) async {
    var member = context.world.members[id]!;
    await member.waitForClose(partnerId);
  },
);

StepDefinitionGeneric thenMemberReceivesConnection =
    then2<String, String, CustomWorld>(
      r'(\S+) receives connection with (\S+)$',
      (id, responderId, context) async {
        var member = context.world.members[id]!;
        await member.waitForConnect(responderId);
      },
    );

StepDefinitionGeneric thenMemberDoesntReceiveConnection =
    then2<String, String, CustomWorld>(
      r"(\S+) doesn't receive connection with (\S+)$",
      (id, responderId, context) async {
        var member = context.world.members[id]!;
        expect(member.connectionStore.connections[responderId], null);
      },
    );

StepDefinitionGeneric thenMemberConnectionStateIs =
    then3<String, String, String, CustomWorld>(
      r"(\S+)'s connection with (\S+) is (\S+)$",
      (id, responderId, connectionState, context) async {
        var member = context.world.members[id]!;
        await member.waitForState(
          responderId,
          parseConnectionState(connectionState),
        );
      },
    );

StepDefinitionGeneric whenConnectionChangesRemoteMediaState =
    when4<String, String, String, String, CustomWorld>(
      r'(\S+) (enables|disables) (audio|video) receiving from (\S+)',
      (id, action, kind, partnerId, context) async {
        var member = context.world.members[id]!;
        var parsedKind = parseMediaKind(kind);

        var connect = member.connectionStore.connections[partnerId]!;
        if (action == 'enables') {
          if (parsedKind.item1 == MediaKind.audio) {
            await connect.enableRemoteAudio();
          } else {
            await connect.enableRemoteVideo();
          }
        } else {
          if (parsedKind.item1 == MediaKind.audio) {
            await connect.disableRemoteAudio();
          } else {
            await connect.disableRemoteVideo();
          }
        }
      },
    );

StepDefinitionGeneric whenMemberReceivesConnection =
    when2<String, String, CustomWorld>(
      r'(\S+) receives connection with (\S+)$',
      (id, responderId, context) async {
        var member = context.world.members[id]!;
        await member.waitForConnect(responderId);
      },
    );
